modules/auxiliary/gather/windows_secrets_dump.rb

```diff
@@ -0,0 +1,1130 @@
+    rescue RubySMB::Dcerpc::Error::WinregError => e
+      vprint_warning("An error occured when getting the default domain name: #{e}")
+      domain = ''
+    end
+    username = "#{domain.encode(Encoding::UTF_8)}\\#{username}" unless domain.nil? || domain.empty?
+    username
+  end
+
+  # Returns Kerberos salt for the current connection if we have the correct information
+  def get_machine_kerberos_salt
+    host = simple.client.default_name
+    return ''.b if host.nil? || host.empty?
+    domain = simple.client.dns_domain_name
+    "#{domain.upcase}host#{host.downcase}.#{domain.downcase}".b
+  end
+
+  def add_parity(byte_str)
+    byte_str.map do |byte|
+      if byte.to_s(2).count('1').odd?
+        (byte << 1) & 0b11111110
+      else
+        (byte << 1) | 0b00000001
+      end
+    end
+  end
+
+  def fix_parity(byte_str)
+    byte_str.map do |byte|
+      t = byte.to_s(2).rjust(8,'0')
+      if t[0,7].count('1').odd?
+        (t[0,7] + '0').to_i(2).chr
+      else
+        (t[0,7] + '1').to_i(2).chr
+      end
+    end
+  end
+
+  def is_weak_des_key?(key)
+    [
+      "\x01\x01\x01\x01\x01\x01\x01\x01",
+      "\xFE\xFE\xFE\xFE\xFE\xFE\xFE\xFE",
+      "\x1F\x1F\x1F\x1F\x0E\x0E\x0E\x0E",
+      "\xE0\xE0\xE0\xE0\xF1\xF1\xF1\xF1",
+      "\x01\xFE\x01\xFE\x01\xFE\x01\xFE",
+      "\xFE\x01\xFE\x01\xFE\x01\xFE\x01",
+      "\x1F\xE0\x1F\xE0\x0E\xF1\x0E\xF1",
+      "\xE0\x1F\xE0\x1F\xF1\x0E\xF1\x0E",
+      "\x01\xE0\x01\xE0\x01\xF1\x01\xF1",
+      "\xE0\x01\xE0\x01\xF1\x01\xF1\x01",
+      "\x1F\xFE\x1F\xFE\x0E\xFE\x0E\xFE",
+      "\xFE\x1F\xFE\x1F\xFE\x0E\xFE\x0E",
+      "\x01\x1F\x01\x1F\x01\x0E\x01\x0E",
+      "\x1F\x01\x1F\x01\x0E\x01\x0E\x01",
+      "\xE0\xFE\xE0\xFE\xF1\xFE\xF1\xFE",
+      "\xFE\xE0\xFE\xE0\xFE\xF1\xFE\xF1"
+    ].include?(key)
+  end
+
+  def aes256_cts_hmac_sha1_96_key(raw_secret, salt)
+    iterations = 4096
+    key_len = 32
+    key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(raw_secret, salt, iterations, key_len)
+    plaintext = "kerberos\x7B\x9B\x5B\x2B\x93\x13\x2B\x93".b
+    rnd_seed = ''.b
+    loop do
+      cipher = OpenSSL::Cipher.new('AES-256-CBC')
+      cipher.encrypt
+      cipher.iv = "\x00".b * 16
+      cipher.key = key
+      ciphertext = cipher.update(plaintext)
+      rnd_seed += ciphertext
+      break unless rnd_seed.size < 32
+      plaintext = ciphertext
+    end
+    rnd_seed.unpack('H*')[0]
+  end
+
+  def aes128_cts_hmac_sha1_96_key(raw_secret, salt)
+    iterations = 4096
+    key_len = 16
+    key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(raw_secret, salt, iterations, key_len)
+    plaintext = "kerberos\x7B\x9B\x5B\x2B\x93\x13\x2B\x93".b
+    rnd_seed = ''.b
+    loop do
+      cipher = OpenSSL::Cipher.new('AES-128-CBC')
+      cipher.encrypt
+      cipher.iv = "\x00".b * 16
+      cipher.key = key
+      ciphertext = cipher.update(plaintext)
+      rnd_seed += ciphertext
+      break unless rnd_seed.size < 16
+      plaintext = ciphertext
+    end
+    rnd_seed.unpack('H*')[0]
+  end
```

Comment:
There's a lot of code repeated in these two functions. It looks like you could combine them and use either a key length parameter of 16 / 32 or an algorithm parameter of aes-128-cbc / aes-256-cbc. AES-128-CBC will always have a 128-bit (16 byte) key and AES-256-CBC will always have a 256-bit (32 byte) key allowing you to safely infer one from the other, though setting the key length from the algorithm could be safer.